### PR TITLE
Add initial Typescript support

### PR DESCRIPTION
**Pull request by aalmazan**
## Description of the change

Adds `tsconfig.json` and `typescript` along with some updates due to `tsc` compiler errors.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

Fixes <your issue here>

### Development

- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development
- [ ] No previous tests unrelated to the changed code fail in development

### Code review

- [ ] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached.
- [ ] At least one reviewer has been requested
- [ ] Changes have been reviewed by at least one other engineer
- [ ] The relevant project board has been selected in Projects to auto-link to this pull request

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -524,27 +534,6 @@ class Anvil {
...
-      const buffer = Buffer.from(data, 'base64')
-      return this._prepareGraphQLBuffer(buffer, options)
-    }
-
-    return {
-      data,
-      filename,
-      mimetype,
-    }
-  }
```
Comment (aalmazan) [inline src/index.js:547]: Got an error inside this method but it seems like we don't use it anymore?

**Comment by newhouse** [comment]
I'm not 100% following what the goal is here?

Is it to write the source code in Typescript and then boil it down to JS? That's what I would think the goal is, right? If that's the case, then there are a few things to fix here:
- The `tsconfig` should not have `emitDeclarationOnly: true`...we want the compiler to output JS files from the TS source files.
- The `tsconfig` should carefully choose the `target` and `lib` values to match with the version of Node that we want the package to support. It's basically like telling it how to babelize things.
- The `/dist` folder should not be checked into github, and should be `.gitignore`'d. The output data should be an artifact of the TS compiling command. However, the `/dist` folder *will* be included in NPM publish, and will end up on Users' machines when they install this package.
- The source files should probably be renamed to `*.ts`.


I can walk you through all this and whatever else I know about Typescript and Node / NPM packages...ping me!

**Comment by aalmazan** [comment]
> I'm not 100% following what the goal is here?

This PR isn't for converting the code base to have typescript support, but to have initial typescript `*.d.ts` files (issue linked above your comment).

> The tsconfig should carefully choose the target and lib values to match with the version of Node that we want the package to support. It's basically like telling it how to babelize things.

Yeah probably needed to pay attention to that part of the config. Switching to:

```
    "target": "es2021",
    "module": "commonjs",
```

> The /dist folder should not be checked into github, and should be .gitignore'd. The output data should be an artifact of the TS compiling command. However, the /dist folder will be included in NPM publish, and will end up on Users' machines when they install this package.

Yes this is on purpose. [From the typescript docs](https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html#editing-the-packagejson), the generated `.d.ts` file path resolution defaults will eventually check `dist/`. IMO I'd prefer that over having `*.d.ts` files beside every `.ts|js` file.

File: src/index.js [hunk 2]
```diff
@@ -1,5 +1,6 @@
 const fs = require('fs')
 
+const { ReadStream } = require('fs')
```
Comment (benogle) [inline src/index.js:3]: was this being used before without import?

File: types/src/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    constructor(options: any);
+    options: any;
+    authHeader: string;
+    hasSetLimiterFromResponse: boolean;
+    limiterSettingInProgress: boolean;
+    rateLimiterSetupPromise: Promise<any>;
+    rateLimiterPromiseResolver: (value: any) => void;
+    _setRateLimiter({ tokens, intervalMs }: {
+        tokens: any;
+        intervalMs: any;
```
Comment (benogle) [inline types/src/index.d.ts:26]: Any easy way to inject types for simple things like this? i.e. this should be an int

File: types/src/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    }): void;
+    limitTokens: any;
+    limitIntervalMs: any;
+    limiter: RateLimiter;
+    createEtchPacket({ variables, responseQuery, mutation }: {
+        variables: any;
+        responseQuery: any;
+        mutation: any;
```
Comment (benogle) [inline types/src/index.d.ts:34]: strings here

**Review by benogle — APPROVED** [review]
Looks good to me! The last thing would be to make sure it can be consumed by a toy typescript project

File: types/src/index.d.ts (hunk 3 above)
Comment (aalmazan) [reply types/src/index.d.ts:26]: Yeah looks like it should generate properly with JSDoc comments. Will add that.

File: package.json [hunk 5]
```diff
@@ -3,6 +3,7 @@
   "version": "2.9.4",
   "description": "Anvil API Client",
   "main": "src/index.js",
+  "types": "./types/index.d.ts",
```
Comment (newhouse) [inline package.json:6]: 👍

**Review by newhouse — APPROVED** [review]

**Comment by newhouse** [comment]
seems like a fair amount of JSDoc'ing could be added to make this really worth it?

**Pull request merged**